stratosphere/ncm/source/ncm_content_meta_database.cpp

```diff
@@ -0,0 +1,584 @@
+        
+        for (auto entry = this->kvs->begin(); entry != this->kvs->end(); entry++) {
+            const auto value = entry->GetValuePointer();
+            const auto header = GetValueHeader(value);
+
+            if (header->content_count == 0) {
+                continue;
+            }
+
+            if (content_ids.GetSize() == 0) {
+                continue;
+            }
+
+            const ContentInfo* content_infos = GetValueContentInfos(value);
+            for (size_t i = 0; i < header->content_count; i++) {
+                const ContentInfo* content_info = &content_infos[i];
+
+                /* Check if any of this entry's content infos matches one of the provided content ids.
+                   If they do, then the content id isn't orphaned. */
+                for (size_t j = 0; j < content_ids.GetSize(); j++) {
+                    const ContentId content_id = content_ids[j];
+
+                    if (content_id == content_info->content_id) {
+                        out_orphaned[j] = false;
+                        break;
+                    }
+                }
+            }
+        }
+
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::Commit() {
+        R_TRY(this->EnsureEnabled());
+        R_TRY(this->kvs->Save());
+        R_TRY(fsdevCommitDevice(this->mount_name));
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::HasContent(sf::Out<bool> out, ContentMetaKey key, ContentId content_id) {
+        const void* value = nullptr;
+        size_t value_size = 0;
+        R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+        const auto header = GetValueHeader(value);
+        const ContentInfo* content_infos = GetValueContentInfos(value);
+
+        if (header->content_count > 0) {
+            for (size_t i = 0; i < header->content_count; i++) {
+                const ContentInfo* content_info = &content_infos[i];
+
+                if (content_id == content_info->content_id) {
+                    out.SetValue(false);
+                    return ResultSuccess();
+                }
+            }
+        }
+
+        out.SetValue(false);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::ListContentMetaInfo(sf::Out<u32> out_entries_written, const sf::OutArray<ContentMetaInfo> &out_meta_info, ContentMetaKey key, u32 start_index) {
+        R_UNLESS(start_index <= std::numeric_limits<s32>::max(), ncm::ResultInvalidOffset());
+        R_TRY(this->EnsureEnabled());
+        
+        const void* value = nullptr;
+        size_t value_size = 0;
+        R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+        const auto header = GetValueHeader(value);
+        const auto content_meta_infos = GetValueContentMetaInfos(value);
+        size_t entries_written = 0;
+
+        if (out_meta_info.GetSize() == 0) {
```

Comment:
I don't understand why you have this if (== 0) return 0 cases all the time.

this code can literally be deleted and the function does the same thing, because of the loop condition check